File: src/main/java/me/sd_master92/customvoting/subjects/VoteTopSign.java
package me.sd_master92.customvoting.subjects;

import me.sd_master92.customvoting.Main;
import me.sd_master92.customvoting.VoteFile;
import me.sd_master92.customvoting.constants.Data;
import me.sd_master92.customvoting.constants.Messages;
import me.sd_master92.customvoting.constants.Voter;
import me.sd_master92.customvoting.database.PlayerTable;
import me.sd_master92.customvoting.extensions.CustomPlaceholders;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.block.Sign;
import org.bukkit.block.Skull;
import org.bukkit.block.data.BlockData;
import org.bukkit.block.data.type.WallSign;
import org.bukkit.entity.Player;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class VoteTopSign
{
    private static final Map<Integer, VoteTopSign> voteTops = new HashMap<>();

    private final Main plugin;
    private final int top;
    private final Location loc;

    public VoteTopSign(Main plugin, int top, Location loc, Player player)
    {
        this.plugin = plugin;
        this.top = top;
        this.loc = loc;

        if (player != null && (top == 0 || !voteTops.containsKey(top)))
        {
            player.sendMessage(ChatColor.GREEN + "Registered Vote Sign #" + (top == 0 ? "title" : top));
        }

        voteTops.put(top, this);
        update();
    }

    public VoteTopSign(Main plugin, int top, Location loc)
    {
        this(plugin, top, loc, null);
    }

    public static VoteTopSign get(Location loc)
    {
        for (int key : voteTops.keySet())
        {
            VoteTopSign voteTop = voteTops.get(key);
            if (voteTop.getLocation().equals(loc))
            {
                return voteTop;
            }
        }
        return null;
    }

    public static void updateAll(Main plugin)
    {
        if (voteTops.isEmpty())
        {
            initialize(plugin);
        }
        new BukkitRunnable()
        {
            @Override
            public void run()
            {
                for (VoteTopSign voteTop : voteTops.values())
                {
                    voteTop.update();
                }
            }
        }.runTaskLater(plugin, 40L);
    }

    private static void initialize(Main plugin)
    {
        Map<String, Location> locations = plugin.getData().getLocations(Data.VOTE_TOP_SIGNS);
        for (String key : locations.keySet())
        {
            Location loc = locations.get(key);
            try
            {
                new VoteTopSign(plugin, Integer.parseInt(key), loc);
            } catch (Exception ignored)
            {
                new VoteTopSign(plugin, 0, loc);
            }
        }
    }

    public Location getLocation()
    {
        return loc;
    }

    private void update()
    {
        if (top == 0)
        {
            updateTitle();
        } else
        {
            updateSign();
        }
    }

    private void updateTitle()
    {
        if (loc.getBlock().getState() instanceof Sign)
        {
            Sign sign = (Sign) loc.getBlock().getState();
            List<String> messages = Messages.VOTE_TOP_SIGNS_TITLE_SIGN.getMessages(plugin);
            for (int i = 0; i < messages.size(); i++)
            {
                sign.setLine(i, messages.get(i));
            }
            sign.update(true);
        }
    }

    private void updateSign()
    {
        if (loc.getBlock().getState() instanceof Sign)
        {
            Sign sign = (Sign) loc.getBlock().getState();
            Voter topVoter = plugin.hasDatabaseConnection() ? PlayerTable.getTopVoter(plugin, top) : VoteFile.getTopVoter(plugin,
                    top);

            if (topVoter != null)
            {
                Location oldLoc = plugin.getData().getLocation(Data.VOTE_TOP_SIGNS + "." + top);
                if (oldLoc != null && !oldLoc.equals(loc))
                {
                    if (oldLoc.getBlock().getState() instanceof Sign)
                    {
                        Sign oldSign = (Sign) oldLoc.getBlock().getState();
                        for (int i = 0; i < 4; i++)
                        {
                            if (i == 1)
                            {
                                oldSign.setLine(i,
                                        Messages.VOTE_TOP_SIGNS_PLAYER_SIGNS_OUTDATED.getMessage(plugin));
                            } else
                            {
                                oldSign.setLine(i, "");
                            }
                        }
                        oldSign.update(true);
                    }
                }

                plugin.getData().setLocation(Data.VOTE_TOP_SIGNS + "." + top, loc);

                Map<String, String> placeholders = new HashMap<>();
                placeholders.put("%NUMBER%", "" + top);
                placeholders.put("%PLAYER%", topVoter.getName());
                placeholders.put("%VOTES%", "" + topVoter.getVotes());
                placeholders.put("%s%", topVoter.getVotes() == 1 ? "" : "s");
                List<String> messages = Messages.VOTE_TOP_SIGNS_PLAYER_SIGNS_FORMAT.getMessages(plugin, placeholders);
                for (int i = 0; i < messages.size(); i++)
                {
                    sign.setLine(i, messages.get(i));
                }
                updateSkull(loc, topVoter.getUuid());
            } else
            {
                for (int i = 0; i < 4; i++)
                {
                    if (i == 1)
                    {
                        sign.setLine(i,
                                Messages.VOTE_TOP_SIGNS_PLAYER_SIGNS_NOT_FOUND.getMessage(plugin));
                    } else
                    {
                        sign.setLine(i, "");
                    }
                }
            }
            sign.update(true);
        }
    }

    public void updateSkull(Location loc, String uuid)
    {
        BlockData blockData = loc.getBlock().getBlockData();
        if (blockData instanceof WallSign)
        {
            WallSign sign = (WallSign) blockData;
            Block block1 = loc.getBlock().getRelative(BlockFace.UP);
            Block block2 = loc.getBlock().getRelative(sign.getFacing().getOppositeFace()).getRelative(BlockFace.UP);
            for (Block block : new Block[]{block1, block2})
            {
                if (block.getState() instanceof Skull)
                {
                    Skull skull = (Skull) block.getState();
                    try
                    {
                        skull.setOwningPlayer(Bukkit.getOfflinePlayer(UUID.fromString(uuid)));
                        skull.update(true);
                    } catch (Exception e)
                    {
                        if (skull.hasOwner())
                        {
                            Material material = skull.getType();
                            BlockFace blockFace = skull.getRotation();

                            block.setType(Material.AIR);
                            new BukkitRunnable()
                            {
                                @Override
                                public void run()
                                {
                                    block.setType(material);
                                    Skull skull = (Skull) block.getState();
                                    if (material == Material.PLAYER_WALL_HEAD)
                                    {
                                        skull.setRotation(sign.getFacing());
                                    } else
                                    {
                                        skull.setRotation(blockFace);
                                    }
                                    skull.update(true);
                                }
                            }.runTaskLater(plugin, 1L);
                        }
                    }
                }
            }
        }
    }

    public void delete(Player player)
    {
        if (plugin.getData().deleteLocation(Data.VOTE_TOP_SIGNS + "." + (top == 0 ? "title" : top)))
        {
            voteTops.remove(top);

            player.sendMessage(ChatColor.RED + "Unregistered Vote Sign #" + (top == 0 ? "title" : top));
        }
    }
}

File: src/main/kotlin/me/sd_master92/customvoting/subjects/VoteTopSign.kt
package me.sd_master92.customvoting.subjects

import com.github.shynixn.mccoroutine.bukkit.launch
import me.sd_master92.core.tasks.TaskTimer
import me.sd_master92.customvoting.CV
import me.sd_master92.customvoting.constants.enumerations.Data
import me.sd_master92.customvoting.constants.enumerations.Message
import me.sd_master92.customvoting.constants.enumerations.PMessage
import me.sd_master92.customvoting.constants.interfaces.Voter
import me.sd_master92.customvoting.getVotesPlaceholders
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.block.BlockFace
import org.bukkit.block.Sign
import org.bukkit.block.Skull
import org.bukkit.block.data.type.WallSign
import org.bukkit.entity.Player
import java.util.*

class VoteTopSign(
    private val plugin: CV,
    private val top: Int,
    val location: Location?,
    player: Player? = null
)
{
    private suspend fun update()
    {
        if (top == 0)
        {
            updateTitle()
        } else
        {
            updateSign()
        }
    }

    private fun updateTitle()
    {
        if (location!!.block.state is Sign)
        {
            val sign = location.block.state as Sign
            val messages = Message.VOTE_TOP_SIGNS_TITLE_SIGN.getMessages(plugin)
            for (i in messages.indices)
            {
                sign.setLine(i, messages[i])
            }
            sign.update(true)
        }
    }

    private suspend fun updateSign()
    {
        if (location!!.block.state is Sign)
        {
            val sign = location.block.state as Sign
            val topVoter = Voter.getTopVoter(plugin, top)
            if (topVoter != null)
            {
                val oldLoc = plugin.data.getLocation(Data.VOTE_TOP_SIGNS.path + "." + top)
                if (oldLoc != null && oldLoc != location)
                {
                    if (oldLoc.block.state is Sign)
                    {
                        val oldSign = oldLoc.block.state as Sign
                        for (i in 0..3)
                        {
                            if (i == 1)
                            {
                                oldSign.setLine(
                                    i,
                                    Message.VOTE_TOP_SIGNS_PLAYER_SIGNS_OUTDATED.getMessage(plugin)
                                )
                            } else
                            {
                                oldSign.setLine(i, "")
                            }
                        }
                        oldSign.update(true)
                    }
                }
                plugin.data.setLocation(Data.VOTE_TOP_SIGNS.path + "." + top, location)
                val placeholders = topVoter.getVotesPlaceholders(plugin)
                placeholders["%NUMBER%"] = "$top"
                for ((i, message) in Message.VOTE_TOP_SIGNS_PLAYER_SIGNS_FORMAT.getMessages(plugin, placeholders)
                    .withIndex())
                {
                    sign.setLine(i, message)
                }
                updateSkull(location, topVoter.getUuid())
            } else
            {
                for (i in 0..3)
                {
                    if (i == 1)
                    {
                        sign.setLine(
                            i,
                            Message.VOTE_TOP_SIGNS_PLAYER_SIGNS_NOT_FOUND.getMessage(plugin)
                        )
                    } else
                    {
                        sign.setLine(i, "")
                    }
                }
            }
            sign.update(true)
        }
    }

    private fun updateSkull(loc: Location, uuid: UUID)
    {
        val sign = loc.block.blockData
        if (sign is WallSign)
        {
            val block1 = loc.block.getRelative(BlockFace.UP)
            val block2 = loc.block.getRelative(sign.facing.oppositeFace)
                .getRelative(BlockFace.UP)
            for (block in arrayOf(block1, block2))
            {
                if (block.state is Skull)
                {
                    val skull = block.state as Skull
                    try
                    {
                        skull.setOwningPlayer(Bukkit.getPlayer(uuid) ?: Bukkit.getOfflinePlayer(uuid))
                        skull.update(true)
                    } catch (e: Exception)
                    {
                        if (skull.hasOwner())
                        {
                            val material = skull.type

                            @Suppress("DEPRECATION")
                            val blockFace = skull.rotation
                            block.type = Material.AIR
                            TaskTimer.delay(plugin)
                            {
                                block.type = material
                                if (material == Material.PLAYER_WALL_HEAD)
                                {
                                    @Suppress("DEPRECATION")
                                    skull.rotation = sign.facing
                                } else
                                {
                                    @Suppress("DEPRECATION")
                                    skull.rotation = blockFace
                                }
                                skull.update(true)
                            }.run()
                        }
                    }
                }
            }
        }
    }

    fun delete(player: Player)
    {
        if (plugin.data.deleteLocation(Data.VOTE_TOP_SIGNS.path + "." + if (top == 0) "title" else top))
        {
            voteTops.remove(top)
            player.sendMessage(PMessage.VOTE_TOP_MESSAGE_SIGN_DELETED_X.with(if (top == 0) "title" else "$top"))
        }
    }

    companion object
    {
        private val voteTops: MutableMap<Int, VoteTopSign> = HashMap()
        operator fun get(loc: Location): VoteTopSign?
        {
            for (key in voteTops.keys)
            {
                val voteTop = voteTops[key]
                if (voteTop!!.location == loc)
                {
                    return voteTop
                }
            }
            return null
        }

        fun updateAll(plugin: CV)
        {
            if (voteTops.isEmpty())
            {
                initialize(plugin)
            }
            TaskTimer.delay(plugin, 40)
            {
                for (voteTop in voteTops.values)
                {
                    plugin.launch {
                        voteTop.update()
                    }
                }
            }.run()
        }

        private fun initialize(plugin: CV)
        {
            val locations = plugin.data.getLocations(Data.VOTE_TOP_SIGNS.path)
            for (key in locations.keys)
            {
                val loc = locations[key]
                try
                {
                    VoteTopSign(plugin, key.toInt(), loc)
                } catch (ignored: Exception)
                {
                    VoteTopSign(plugin, 0, loc)
                }
            }
        }
    }

    init
    {
        if (player != null && (top == 0 || !voteTops.containsKey(top)))
        {
            player.sendMessage(PMessage.VOTE_TOP_MESSAGE_SIGN_CREATED_X.with(if (top == 0) "title" else "$top"))
        }
        voteTops[top] = this
        plugin.launch {
            update()
        }
    }
}